### PR TITLE
Audit fix MAT1133 MAT1140, MAT1143, MAT1142,  MAT1128

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

File: contracts/staking/stakeManager/StakeManager.sol [hunk 1]
```diff
@@ -343,6 +349,7 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
     // slashing and jail interface
     function restake(uint256 validatorId, uint256 amount, bool stakeRewards)
         public
+        onlyWhenUnlocked,
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:352]: @AyushKaul check if multiple modifiers work with comma or space

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 2]
```diff
@@ -125,6 +127,7 @@ contract ValidatorShare is IValidatorShare {
         activeAmount = activeAmount.sub(_amount);
 
         amountStaked[msg.sender] = 0;
+        //TODO: check if exists @ayush then update amount, reWrite new withdraw delay
```
Comment (0xAshish) [inline contracts/staking/validatorShare/ValidatorShare.sol:130]: MAT-1142

File: contracts/staking/stakeManager/StakeManager.sol [hunk 3]
```diff
@@ -31,7 +31,10 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
     constructor() public Lockable(address(0x0)) {}
 
     // TopUp heimdall fee
-    function topUpForFee(uint256 validatorId, uint256 heimdallFee) public {
+    function topUpForFee(uint256 validatorId, uint256 heimdallFee)
+        public
+        onlyWhenUnlocked
+    {
```
Comment (0xAshish) [inline contracts/staking/stakeManager/StakeManager.sol:37]: @AyushKaul add a check if the validator is active or not, `require(isValidator(validator))`.

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 4]
```diff
@@ -90,6 +90,8 @@ contract ValidatorShare is IValidatorShare {
 
     function buyVoucher(uint256 _amount) public onlyWhenUnlocked {
         uint256 share = _amount.mul(100).div(exchangeRate());
+        require(share > 0, "Insufficient amount to buy share ");
```
Comment (0xAshish) [inline contracts/staking/validatorShare/ValidatorShare.sol:93]: Also, Please add a check `require(delegators[msg.sender].amount==0)` for MAT1142

Comment (AyushKaul) [reply contracts/staking/validatorShare/ValidatorShare.sol:93]: done

File: contracts/staking/stakeManager/StakeManager.sol (hunk 3 above)
Comment (AyushKaul) [reply contracts/staking/stakeManager/StakeManager.sol:37]: done

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 2 above)
Comment (AyushKaul) [reply contracts/staking/validatorShare/ValidatorShare.sol:131]: removed

File: contracts/staking/stakeManager/StakeManager.sol (hunk 1 above)
Comment (AyushKaul) [reply contracts/staking/stakeManager/StakeManager.sol]: fixed

**Comment by AyushKaul** [comment]
@0xAshish can you quickly take a look and add suggestions before I add tests

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 5]
```diff
@@ -90,6 +90,9 @@ contract ValidatorShare is IValidatorShare {
...
+        require(delegators[msg.sender].amount ==0, "Delegator has no rewards left");
```
Comment (0xAshish) [inline contracts/staking/validatorShare/ValidatorShare.sol:94]: Please update the require msg to something like "Ongoing exit" or something like that.

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 2 above)
Comment (0xAshish) [reply contracts/staking/validatorShare/ValidatorShare.sol:131]: removed this todo?

File: contracts/staking/validatorShare/ValidatorShare.sol [hunk 6]
```diff
@@ -89,6 +89,8 @@ contract ValidatorShare is IValidatorShare {
 
     function buyVoucher(uint256 _amount) public onlyWhenUnlocked {
         uint256 share = _amount.mul(100).div(exchangeRate());
+        require(share > 0, "Insufficient amount to buy share ");
```
Comment (AyushKaul) [inline contracts/staking/validatorShare/ValidatorShare.sol:93]: @0xAshish this check doesn't seem right, it should be `_amount > 0`

Comment (0xAshish) [reply contracts/staking/validatorShare/ValidatorShare.sol:93]: This is fine, id amount is zero then share count will be by default zero.

File: contracts/staking/validatorShare/ValidatorShare.sol (hunk 5 above)
Comment (0xAshish) [reply contracts/staking/validatorShare/ValidatorShare.sol]: @AyushKaul ☝️

**Comment by AyushKaul** [comment]
@0xAshish can you please review

File: test/staking/Validator.test.js [hunk 7]
```diff
@@ -276,6 +296,34 @@ contract('ValidatorShare', async function (accounts) {
     let logs = logDecoder.decodeLogs(result.receipt.rawLogs)
     assertBigNumberEquality(logs[0].args.value, web3.utils.toWei('100'))
   })
+
+  it('Buy -> sell -> buy -> expect error', async function () {
```
Comment (0xAshish) [inline test/staking/Validator.test.js:300]: Please add better `it` description.

Comment (AyushKaul) [reply test/staking/Validator.test.js:300]: okay I'll do

**Comment by AyushKaul** [comment]
@0xAshish  changes are done. We can merge now :smile:

**Pull request merged**